### ISSUE TITLE
pywin32_postinstall.py not robust to creating multiple environments

### ISSUE THREAD

**Issue by zarquon5** [mhammond/pywin32#2032]
Python3.9
pywin32 302

I have a situation where I am installing a second virtual environment with pywin32 on a given host, so I can cut over to a new service instance from a set of services currently running, after the new environment and software is set up.  

After building the new environment, I am trying to run pywin32_postinstall.py _in_ that environment, which tries to (now overwrite) the files in the windows folder (in my case python39.dll, pythoncom39.dll, etc) previously copied there from running the same script in my first environment, but this fails because they're currently in use by a win32 python service running in that first environment.  

And, because it raises the exception, the rest of the post install code doesn't run, so the pythonservice.exe in the 2nd environment is then not registered.  So, if I try to only run pywin32_postinstall.py iff the files aren't present in the windows root folder, then when I try to install my new service, I get:
```
File "C:\MyReleaseFolder\AppSrvService.py", line 252, in <module>
    win32serviceutil.HandleCommandLine(AppSrvService)
  File "c:\MyNewEnvironment\lib\site-packages\win32\lib\win32serviceutil.py", line 833, in HandleCommandLine
    InstallService(
  File "c:\MyNewEnvironment\lib\site-packages\win32\lib\win32serviceutil.py", line 209, in InstallService
    exeName = '"%s"' % LocatePythonServiceExe(
  File "c:\MyNewEnvironment\lib\site-packages\win32\lib\win32serviceutil.py", line 56, in LocatePythonServiceExe
    raise error(msg)
RuntimeError: c:\MyNewEnvironment\Scripts\pythonservice.exe is not correctly registered
Please locate and run c:\MyNewEnvironment\Scripts\pythonservice.exe, and it will self-register
Then run this service registration process again.
```

I don't want to take down the running service, as the work done to prep the new environment after pywin32 is installed is non-trivial, and would result in non-trivial downtime for our service.  Would it be possible to change the script to first check whether the files are already in place before copying them, or making it gracefully continue if the overwrite fails?

i.e.
```            for fname in files:
                base = os.path.basename(fname)
                dst = os.path.join(dest_dir, base)
                if not os.path.exists(dst):
                    CopyTo("installing %s" % base, fname, dst)
                    if verbose:
                        print("Copied %s to %s" % (base, dst))
                else:
                    print("%s already present" % (dst))
```

(Or is there another way for me to get this done without replicating this script?)
Thanks in advance,
/Will Sadkin

**Comment by mhammond** [comment]
Checking if the files are already in place is going to break people upgrading pywin32.

The main point of this script is to set up some "global" things - can't you just not run it in that scenario? What breaks there?

**Comment by zarquon5** [comment]
I don't understand why trying to overwrite the python[com]<ver>.dlls would break upgrades of pywin32; if the same version of python is in use for the different versions, this doesn't change, and if upgrading python, then they're have different names... (Or is the above code used more generally for other things that might have the same filename but different versions?)  What am I missing?  

As far as "global things", I want to be able to run different versions of code and packages (including possibly pywin32?) in my pythonservice instances, often with the same version of python, and so I've been creating separate conda virtual environments to manage the package configurations.  But with the above problem, upgrading from one version to the other requires that effectively no service be running while I do it, because it's trying to overwrite a file that's in use.

**Comment by zarquon5** [comment]
Also, FTR, I've found running pythonservices from virtual environments to be a tricky thing to do, since windows is not aware of them.  I've resorted to effectively writing my own "activate" from within the service, per this recipe:
https://stackoverflow.com/questions/34696815/using-pythonservice-exe-to-host-python-service-while-using-virtualenv

**Comment by mhammond** [comment]
The issue with upgrading is that a new version of pywin32 will have updated pywintypesXX.dll, but the copy from the older pywin32 will exist.

The copying of these files is main done exactly for things like services, and, say COM objects, which really are designed to be "global" on the machine. I don't see how it really makes much sense for a service to be hosted in a virtual env.

If you really want these environments to possibly have different versions of pywin32, then running that postinstall script doesn't make any sense - as you note, it tries to copy pywintypesXX.dll into a global location, but different pywin32 versions have different, often incompatible, versions of that file. It sounds to me like you need to decide whether you want pywin32 to be local to a virtual-env, in which case you can't run the postinstall script, or one global version, in which case you should install it into your python environment (ie, not a virtualenv) and run it once.

**Comment by mhammond** [comment]
I've tried to clarify some of this in 0906bd88.

**Comment by zarquon5** [comment]
Hi Mark, 

My use case is that I need to be able to set up a new service instance on an existing fielded service machine, potentially having newer packages and/or additional functionality, and I want to both delay turning off the old service until everything is ready for cut-over (which can take some time, as there can be artifacts that the service can need to rebuild in light of its new features, that take non-trivial time to create), and also allow for the possibility of rollback to the earlier version if for some reason the new service has some unexpected problem during startup.
 
The way I managed to get this to work for me was to, during installation, place all of the required DLLs into the Library\bin folder of the venv, then put (or add to) a VENV_PATH variable in the system environment containing that path, and then add this variable to/ensure its presence in the system Path variable.  I then duplicated some of the (non-encapsulated) code out of the pywin32_postinstall module to set up the gen_py folder, and then imported pywin32_postinstall and called its RegisterCOMObjects() function.

This strategy seemed to work for my needs, in that it allowed an existing running service to run with whatever it previously found along the VENV_PATH, while also allowing a new service, potentially with a different set of supporting DLLs to be installed and start, and avoided the problem of trying to overwrite a file in the file system that's currently in use by the current service. 

But it doesn't solve the problem of having two potentially incompatible versions of pywintypesXX.dll on my VENV_PATH path, and if, for some reason, I have to roll back my upgrade I'd presumably have to remove the new venv's Library\bin from the VENV_PATH first so that the old service can restart, because otherwise the old service would now find the new (potentially incompatible) pywintypes dll first on that path.  (This still feels like a versioning problem to me, as it seems imprudent to have different versions of a dll that share the same name and yet aren't backward-compatible with each other; am I missing something?)

Regards,
/Will Sadkin
